### PR TITLE
Juju logo should link to appropriate place

### LINKED ISSUE #2729: Click to the Juju logo breaks gijoe
It should not just point to / when not in gisf, it should use the state system instead.

### PR DESCRIPTION
**Pull request by hatched**
This branch updates the header logo so that it links to the appropriate place depending on its environment. In GISF we want the Juju logo to navigate to / and in GIJOE we want it to navigate to /u/{username}.

As a driveby two updates were made to the state system.
- A `reset` method was added which sets every active state object to null
- `generatePath` method now takes an arbitrary state object.

Fixes #2729

### CODE REVIEW

**Comment by hatched** [comment]
#### To QA

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7093/

File: jujugui/static/gui/src/app/components/header-logo/header-logo.js [hunk 1]
```diff
@@ -27,15 +27,43 @@ YUI.add('header-logo', function() {
...
+            {this._svg()}
+          </a>);
+      }
       return (
-        <a href="/"
-          role="button" title="Go to user profile">
-          <juju.components.SvgIcon name="juju-logo"
-            className="svg-icon"
-            width="90" height="35" />
+        <a onClick={this._showProfile}
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/header-logo/header-logo.js]: Use `generatePath`'s new ability here to add an `href` attribute pointing to the profile.

File: jujugui/static/gui/src/app/components/header-logo/test-header-logo.js [hunk 2]
```diff
@@ -30,17 +30,46 @@ describe('HeaderLogo', function() {
...
+    const expected = (
+      <a onClick={showProfile} role="button" title="Home">
+        <juju.components.SvgIcon name="juju-logo"
+          className="svg-icon"
+          width="90" height="35" />
+      </a>);
+    expect(output).toEqualJSX(expected);
+  });
+
+  it('calls showProfil on click in gisf', () => {
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/header-logo/test-header-logo.js]: `showProfile`

File: jujugui/static/gui/src/app/state/state.js [hunk 3]
```diff
@@ -573,15 +573,18 @@ const State = class State {
 
   /**
     Takes the existing app state and generates the path.
+    @param {Object} stateObj An optional argument to generate a path from using
```
Comment (kadams54) [inline jujugui/static/gui/src/app/state/state.js]: 😄

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7094/

**Comment by kadams54** [comment]
QA is OK.

**Comment by hatched** [comment]
:shipit: 
Thanks

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**